### PR TITLE
Add Secure cookie flag option to session cookie

### PR DESCRIPTION
**Pull request by SteveArr**
The session cookie needs the ability to have the 'Secure' flag set.  This PR adds support for always use secure flag, never use secure flag and use secure flag only if initial request is SSL.

Included unit tests for Secure flag as well as HTTPOnly flag.

### CODE REVIEW

**Comment by msotclas** [comment]
Hi __@SteveArr__, I'm your friendly neighborhood Microsoft Open Technologies, Inc. Pull Request Bot (You can call me MSOTBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla2.msopentech.com.
    </span>

TTYL, MSOTBOT;

**Comment by msotclas** [comment]
__@SteveArr__, Thanks for signing the contribution license agreement so quickly! Actual humans will now validate the agreement and then evaluate the PR.
<br />Thanks, MSOTBOT;

**Comment by Eilon** [comment]
@blowdart can you have a look?

**Comment by blowdart** [comment]
My usual naming issues appear

Https != Secure (We should look at CookieOptions as well to fix this)

There's no other problem aside from that.

File: src/Microsoft.AspNet.Session/SessionDefaults.cs [hunk 1]
```diff
@@ -9,5 +9,7 @@ public static class SessionDefaults
     {
         public static string CookieName = ".AspNet.Session";
```
Comment (Eilon) [inline src/Microsoft.AspNet.Session/SessionDefaults.cs]: @Tratcher are all these suppose to be `readonly`?

**Comment by blowdart** [comment]
There is a side effect here, in that if you switch protocols you will loose session if you're switching from HTTPS to HTTP. I don't consider this a bad thing, other opinions may vary.

File: src/Microsoft.AspNet.Session/SessionDefaults.cs (hunk 1 above)
Comment (Tratcher) [inline src/Microsoft.AspNet.Session/SessionDefaults.cs]: Yes

File: src/Microsoft.AspNet.Session/CookieSecureOption.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
```
Comment (Eilon) [inline src/Microsoft.AspNet.Session/CookieSecureOption.cs]: BTW we just changed the copyright noticed to:

`Copyright (c) .NET Foundation. All rights reserved.`

File: src/Microsoft.AspNet.Session/CookieSecureOption.cs [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNet.Session
+{
+    /// <summary>
+    /// Determines how the identity cookie's security property is set.
```
Comment (Eilon) [inline src/Microsoft.AspNet.Session/CookieSecureOption.cs]: BTW a lot of this new code mentions "identity" yet this is all about session. Is that deliberate?

**Comment by blowdart** [comment]
This would probably be better addressed as part of abstractions, where there is already an open issue on the topic. https://github.com/aspnet/HttpAbstractions/issues/42

File: src/Microsoft.AspNet.Session/CookieSecureOption.cs (hunk 3 above)
Comment (Tratcher) [inline src/Microsoft.AspNet.Session/CookieSecureOption.cs]: This is a direct copy from the cookie auth middleware:
https://github.com/aspnet/Security/blob/dev/src/Microsoft.AspNet.Authentication.Cookies/CookieSecureOption.cs

Comment (blowdart) [inline src/Microsoft.AspNet.Session/CookieSecureOption.cs]: We should rename that too then

**Comment by SteveArr** [comment]
After reading your comments, here is the change I propose:

Take CookieSecureOption.cs and move it up into Aspnet/HttpAbstractions like @blowdart is recommending.  Update code comments appropriately.

Update this pull request to use that relocated CookieSecureOption.cs

Update Aspnet/Security/Microsoft.AspNet.Authentication.Cookies to also use the newly located CookieSecureOption.cs to reduce code duplication.

Question:  I am unfamiliar with the process for updating code in three different projects.  Do I have to submit a PR to HTTPAbstractions first and let that get in before making the changes to Session and Security?

**Comment by Tratcher** [comment]
You can submit concurrent PRs in multiple repos, just and links to each related PR in the comments.

Do you know how to use the global.json file to compile multiple repos locally?

**Comment by SteveArr** [comment]
@Tratcher ,  I am unaware of how to use the global.json file to compile multiple repos locally.  Can you forward me some information or pointers?

**Comment by Tratcher** [comment]
In Session modify the global.json add a relative reference to HttpAbstractions. This way your local changes will be used rather than the nuget packages.
```
{
    "projects": [ "src", "..\\HttpAbstractions\\src" ]
}
```

Do not check in this change, it is only for local development.

**Comment by SteveArr** [comment]
Got it.  I will work on this tonight when I get home.

**Comment by muratg** [comment]
@Tratcher Could you look at this one (and the corresponding HttpAbstractions one)?

**Comment by Tratcher** [comment]
@SteveArr Can you update your CLA as described above?

**Comment by muratg** [comment]
@SteveArr Are you still looking into this one?

**Comment by muratg** [comment]
Closing this one, we'll do https://github.com/aspnet/httpAbstractions/issues/42 instead.

**Pull request closed without merging**